proof/refine/ARM/Schedule_R.thy

```diff
@@ -4862,136 +4869,438 @@ lemma refillBudgetCheck_corres:
+              and pspace_distinct)
+             (valid_tcbs' and valid_queues and valid_queues' and valid_release_queue_iff)
+             check_domain_time
+             checkDomainTime"
+  apply (clarsimp simp: check_domain_time_def checkDomainTime_def)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF isCurDomainExpired_corres])
+      apply (rule corres_when)
+       apply simp
+      apply (rule corres_split[OF setReprogramTimer_corres])
+        apply (rule rescheduleRequired_corres)
+       apply (wpsimp wp: hoare_drop_imps
+                   simp: isCurDomainExpired_def)+
+  done
+
+crunches refill_budget_check_round_robin
+  for sc_at[wp]: "sc_at sc_ptr"
+  and valid_objs[wp]: valid_objs
+  and pspace_aligned[wp]: pspace_aligned
+  and pspace_distinct[wp]: pspace_distinct
+
+lemma commitTime_corres:
+  "corres dc (\<lambda>s. sc_at (cur_sc s) s \<and> valid_objs s \<and> pspace_aligned s \<and> pspace_distinct s
+                  \<and> (cur_sc_active s \<longrightarrow> valid_refills (cur_sc s) s))
+             (\<lambda>s'. is_active_sc' (ksCurSc s') s' \<longrightarrow> valid_refills' (ksCurSc s') s')
+             commit_time
+             commitTime"
   supply if_split[split del]
-  apply (clarsimp simp: Schedule_A.schedule_def Thread_H.schedule_def)
-  sorry (* schedule_corres *) (*
-  apply (subst thread_get_test)
-  apply (subst thread_get_comm)
-  apply (subst schact_bind_inside)
+  apply (rule_tac Q="\<lambda>s'. sc_at' (ksCurSc s') s'" in corres_cross_add_guard)
+   apply (fastforce intro: sc_at_cross simp: state_relation_def)
+  apply (clarsimp simp: commit_time_def commitTime_def liftM_def)
+  apply (rule corres_split'[rotated 2, OF gets_sp getCurSc_sp])
+   apply (corressimp corres: getCurSc_corres)
+  apply clarsimp
+  apply (rule corres_split'[rotated 2, OF get_sched_context_sp get_sc_sp'])
+   apply (corressimp corres: get_sc_corres)
+  apply (rule corres_split'[rotated, where r'=dc])
+     apply (rule setConsumedTime_corres)
+     apply simp
+    apply wpsimp
+   apply wpsimp
+  apply (clarsimp simp: when_def)
+  apply (rule corres_if_split; fastforce?)
+   apply (clarsimp simp: sc_relation_def active_sc_def)
+  apply (rule corres_split'[rotated 2, OF gets_sp getConsumedTime_sp])
+   apply corressimp
+  apply clarsimp
+  apply (rename_tac consumed)
+  apply (rule_tac Q="\<lambda>_ s. sc_at (cur_sc s) s \<and> csc = cur_sc s"
+              and Q'="\<lambda>_ s'. sc_at' (ksCurSc s') s' \<and> csc = ksCurSc s'"
+              and r'=dc
+               in corres_split'[rotated])
+     apply (clarsimp simp: updateSchedContext_def)
+     apply (rule corres_symb_exec_r[rotated, OF get_sc_sp'])
+       apply wpsimp
+      apply wpsimp
+     apply (rename_tac sc')
+     apply (prop_tac "(\<lambda>sc. sc\<lparr>sc_consumed := sc_consumed sc + consumed\<rparr>)
+                      = sc_consumed_update (\<lambda>c. c + consumed)")
```

Comment:
Nothing to do with this PR, but it's a bit disturbing that the machinery needs this manually added.

Reply:
Indeed